File: Algorithms/src/andy/level1/FailureLate.java
package andy.level1;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

//프로그래머스 Level 1,실패율
class Node {
    public int no;
    public float fail;
    
    Node(int i,float f) {
        no = i;
        fail = f;
    }
}


class NodeCompare implements Comparator<Node> {
	
	public int compare(Node a, Node b) {
		if(a.fail < b.fail)
			return 1;
		else if(a.fail > b.fail)
			return -1;
		return 0;
	}
}


public class FailureLate {
    public int[] solution(int N, int[] stages) {
        List<Node> fail = new ArrayList<Node>();
        int cnt = stages.length;
        int[] answer = new int[N];
        int[] arr = new int[N+2];
        Arrays.fill(arr, 1);   
        for(int d:stages) arr[d]++;
        for(int i=1;i<=N;i++) {
            fail.add(new Node(i,arr[i]/cnt));
            cnt -= arr[i];
        }
        
        Collections.sort(fail, new NodeCompare());
        
        int i=0;
        for(Node d:fail) {
        	answer[i++] = d.no;
        }
        
        
        return answer;
    }
}

File: Andy/src/level1/javascript/FailureLate.js
// 프로그래머스 Level 1,실패율
function solution(N, stages) {
    var answer = [];   
    var cnt = stages.length;
    var arr = Array.from({length:N+2},() =>0);
    stages.forEach( (d) => { arr[d]++; });    
    var fail = [];
    for(let i=1;i<=N;i++) {
        fail[i] = { 'no':i, 'f':arr[i]/cnt };
        cnt -= arr[i];
    }    
    fail.sort( (a,b) => { return b.f-a.f }).forEach( (d) => { answer.push(d.no) });  
    return answer;
}
